### PR TITLE
Modernized code with clang-tidy [PCL-2731]

### PR DESCRIPTION
**Pull request by gnawme**
Initial modernization of code base using `clang-tidy` and an innocuous subset of its `modernize` checks:

- [ ] modernize-avoid-bind
- [ ] modernize-avoid-c-arrays
- [ ] modernize-concat-nested-namespaces
- [x] modernize-deprecated-headers
- [ ] modernize-deprecated-ios-base-aliases
- [ ] modernize-loop-convert
- [ ] modernize-make-shared
- [ ] modernize-make-unique
- [ ] modernize-pass-by-value
- [ ] modernize-raw-string-literal
- [x] modernize-redundant-void-arg
- [ ] modernize-replace-auto-ptr
- [ ] modernize-replace-random-shuffle
- [ ] modernize-return-braced-init-list
- [ ] modernize-shrink-to-fit
- [ ] modernize-unary-static-assert
- [ ] modernize-use-auto
- [ ] modernize-use-bool-literals
- [ ] modernize-use-default-member-init
- [ ] modernize-use-emplace
- [x] modernize-use-equals-default
- [x] modernize-use-equals-delete
- [ ] modernize-use-nodiscard
- [ ] modernize-use-noexcept
- [x] modernize-use-nullptr
- [x] modernize-use-override
- [ ] modernize-use-trailing-return-type
- [ ] modernize-use-transparent-functors
- [ ] modernize-use-uncaught-exceptions
- [x] modernize-use-using

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -8,8 +8,10 @@ if(POLICY CMP0074)
 endif()
 
 # Set target C++ standard and required compiler features
-set(CMAKE_CXX_STANDARD 14 CACHE STRING "The target C++ standard. PCL requires C++14 or higher.")
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+set(CMAKE_CXX_STANDARD 17 CACHE STRING "The target C++ standard. PCL requires C++17 or higher.")
```
Comment (kunaltyagi) [inline CMakeLists.txt:12]: Why did you change the C++ standard?

**Comment by kunaltyagi** [comment]
Could I request you to help integrate clang-tidy in the CI before adding the improvements to prevent a backslide later on?

File: CMakeLists.txt (hunk 1 above)
Comment (gnawme) [reply CMakeLists.txt:12]: Curiosity. 

You seemed to indicate in [this issue](https://github.com/PointCloudLibrary/pcl/issues/4083) that you were OK with moving to C++17, so I thought I'd experiment and see how CI would handle it.

**Comment by gnawme** [comment]
> Could I request you to help integrate clang-tidy in the CI before adding the improvements to prevent a backslide later on?

@kunaltyagi If there's a way to run `clang-tidy` as a last-stage check before merging, sure. 

As something to run with each CI build, however, it makes CI much longer without adding much value. Policy at places I've worked is for the developer to run `clang-tidy` when they're getting close to be able to merge.

**Comment by kunaltyagi** [comment]
Does it take long to run clang-tidy? (How long did it take to run on PCL?)

**Comment by gnawme** [comment]
> Does it take long to run clang-tidy? (How long did it take to run on PCL?)

Try this:

- Add `set(CMAKE_EXPORT_COMPILE_COMMANDS ON)` to `CMakeLists.txt`
- Run `cmake .`
- Make sure `compile_commands.json` is present.
- Assuming you have `clang-tidy` installed, run (for example) `run-clang-tidy -header-filter='.*' -checks=".*modernize-use-nullptr" -fix`

**Comment by kunaltyagi** [comment]
`clang-tidy` on my system takes 20 minutes, with a full compile (including cuda modules) requiring 45 minutes.

@PointCloudLibrary/maintainers What about trying GitHub actions for this?

**Comment by SergioRAgostinho** [comment]
No opposition from my side.

**Comment by kunaltyagi** [comment]
@gnawme Could you please add a clang-tidy action to this PR on the lines of the format CI:
* clone in a docker
* run clang-format
* diff to see if there's a regression
* show diff else pass

https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions

**Comment by gnawme** [comment]
> @gnawme Could you please add a clang-tidy action to this PR on the lines of the format CI:
> 
> * clone in a docker
> * run clang-format
> * diff to see if there's a regression
> * show diff else pass
> 
> https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions

@kunaltyagi Add this sequence to `formatting.yaml`? OK, I'll take a look.

**Comment by gnawme** [comment]
> > @gnawme Could you please add a clang-tidy action to this PR on the lines of the format CI:
> > 
> > * clone in a docker
> > * run clang-format
> > * diff to see if there's a regression
> > * show diff else pass
> > 
> > https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions
> 
> @kunaltyagi Add this sequence to `formatting.yaml`? OK, I'll take a look.

This will be in `.ci/azure-pipelines/tidying.yaml`; how do I test this?
=====
```
jobs:
  - job: tidying
    displayName: Modernize C++
    pool:
      vmImage: 'ubuntu-latest'
    container: tdy
    steps:
      - checkout: self
        # find the commit hash on a quick non-forced update too
        fetchDepth: 10
      - script: $(which run-clang-tidy) -header-filter='.*' -checks='-*,modernize-deprecated-headers,modernize-redundant-void-arg,modernize-replace-random-shuffle,modernize-use-equals-default,modernize-use-equals-delete,modernize-use-nullptr,modernize-use-override,modernize-use-using' -fix
        displayName: 'Run clang-tidy'
      - script: git diff > formatting.patch
        displayName: 'Compute diff'
      - script: cat formatting.patch
        displayName: 'Show diff'
      - task: CopyFiles@2
        inputs:
          sourceFolder: '$(Build.SourcesDirectory)'
          contents: 'formatting.patch'
          targetFolder: '$(Build.ArtifactStagingDirectory)'
        displayName: 'Copy diff to staging directory'
      - task: PublishBuildArtifacts@1
        inputs:
          pathtoPublish: '$(Build.ArtifactStagingDirectory)'
          artifactName: formatting
        displayName: 'Publish diff'
      - script: "[ ! -s formatting.patch ]"
        displayName: 'Set job exit status'
```

**Comment by kunaltyagi** [comment]
> how do I test this?

I don't know. Does the file needs to be in master for this? @shrijitsingh99

**Comment by shrijitsingh99** [comment]
> how do I test this?

Just setup Github Actions on your fork. That should work. Nothing to be done on PCL side to test.

> Could I request you to help integrate clang-tidy in the CI before adding the improvements to prevent a backslide later on?
I would recommend looking into #4004 . The idea was to first integrate it into CMake or provide some tooling for running clang-tidy then integrate it. As I remember the blocker was CMake version since newer versions had tidy support.

Without providing tooling for contributors to run `clang-tidy` easily, adding a check to CI won't be that useful IMO.

**Comment by gnawme** [comment]
> I would also recommend looking into #4004. The idea was to first integrate it into CMake or provide some tooling for running clang-tidy then integrate it. As I remember the blocker was CMake version since newer versions had tidy support.

Older CMake versions are also blocking the transition to C++17 or later, since they don't recognize `CMAKE_CXX_STANDARD 17`. Are there plans to remedy that going forward?

> Without providing tooling for contributors to run `clang-tidy` easily, adding a check to CI won't be that useful IMO.

Yes, `clang-tidy` was left as a developer task at places I worked that used `clang-tidy`. (There was a bash script in the repo's `.ci` directory that wasn't actually run by CI.)

**Comment by shrijitsingh99** [comment]
> Are there plans to remedy that going forward?

Have opened an issue #4004, we can continue the discussion there.

> Yes, `clang-tidy` was left as a developer task at places I worked that used `clang-tidy`. (There was a bash script in the repo's `.ci` directory that wasn't actually run by CI.)

Yeah, for now at least we can also provide a bash script. Maybe using the commands you are currently running for `clang-tidy`.

**Comment by gnawme** [comment]
@kunaltyagi once these changes are approved, I think the affected files could use a `clang-format` pass.

**Comment by kunaltyagi** [comment]
> I think the affected files could use a `clang-format` pass.

We've been rolling out clang-format bit by bit (since the patch can be monstrous as well has mess up the comments and some code). Which module do you think should be prioritised? PR by modifying the whitelist in `.dev` and running format 2 times would suffice though it might not get reviewed soon (GSoC pressure)

File: .ci/azure-pipelines/tidying.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+jobs:
```
Comment (kunaltyagi) [inline .ci/azure-pipelines/tidying.yaml:1]: @shrijitsingh99 Could you please check this? You've played with GitHub actions more than me. This feels like Azure format to me.

Also, @gnawme if you could run this on your fork and link the results that'd be great

**Comment by gnawme** [comment]
> > I think the affected files could use a `clang-format` pass.
> 
> We've been rolling out clang-format bit by bit (since the patch can be monstrous as well has mess up the comments and some code). Which module do you think should be prioritised? PR by modifying the whitelist in `.dev` and running format 2 times would suffice though it might not get reviewed soon (GSoC pressure)

These are the directories where the `clang-tidy` changes occurred in `1b90309b9`:

- common
- features
- filters
- geometry
- io
- kdtree
- keypoints
- ml
- octree
- outofcore
- people
- recognition
- registration
- sample_consensus
- search
- segmentation
- stereo
- surface
- tools
- tracking
- visualization

Which will give most bang for the buck if they had `clang-format` run on them?

**Comment by kunaltyagi** [comment]
Lines of code affected in each module would be a better indicator. You can get per directory stats (in alphabetical order) using:

```bash
ls -d ./* | xargs -n1 bash -c 'git diff --numstat master -- $@ | python -c "import fileinput as f; print(sum([int(line.strip()[0]) for line in f.input()]))"' {}
```

The largest one should be the priority

**Comment by SergioRAgostinho** [comment]
I would say following dependency graph should be top priority. The changes you make in the "base modules" are more likely to affect more parts of the code.

**Comment by gnawme** [comment]
> I would say following dependency graph should be top priority. The changes you make in the "base modules" are more likely to affect more parts of the code.

In lines of code and connections in the dependency graph, looks like:

- io (11764)

- visualization (6600)

- features (6323)

- surface (7418)

`recognition` has high LOC, but doesn't look as central in the dependency graph.

File: .ci/azure-pipelines/tidying.yaml (hunk 2 above)
Comment (shrijitsingh99) [reply .ci/azure-pipelines/tidying.yaml:1]: Yeah, this is an azure format plus github actions are meant to be in `.github/workflow`

Comment (gnawme) [reply .ci/azure-pipelines/tidying.yaml:1]: Probably moot, since @kunaltyagi stated a preference for integrating `clang-tidy` into CMake versions that support it directly?

But yes, I think I copied an Azure script. I'll revisit this.

**Comment by gnawme** [comment]
@kunaltyagi @SergioRAgostinho I ran `clang-format` on `common`, `kdtree`, and `octree`, net 1281 LOC changed (include `clang-tidy` changes). Shall I run `clang-format` on another module or two?

**Comment by SergioRAgostinho** [comment]
> > I would say following dependency graph should be top priority. The changes you make in the "base modules" are more likely to affect more parts of the code.
> 
> @SergioRAgostinho Wow, your graph is a lot cleaner and simpler than the one produced by `cmake --graphviz=graph`. What did you use to produce it?

I didn't, @divmadan did :)


> @kunaltyagi @SergioRAgostinho I ran `clang-format` on `common`, `kdtree`, and `octree`, net 1281 LOC changed (including `clang-tidy` changes). Shall I run `clang-format` on another module or two?

My preferred approach is to submit a PR per tool per module e.g. pass cmake tidy over common open PR, once merged enable cmake format for common, continue for other modules. We just had a PR of equivalent magnitude in LoC changes, a "simple find replace", and it was a nightmare to review.

File: .ci/azure-pipelines/tidying.yaml (hunk 2 above)
Comment (kunaltyagi) [reply .ci/azure-pipelines/tidying.yaml:1]: > Probably moot, ... integrating clang-tidy into CMake versions that support it directly

We can switch to a pure CMake version, but adding the GitHub Action for compile-commands + bash should not be hard. We can make an issue to track the cmake integration as a separate actionable item (which based on your previous comment, it is)

**Comment by gnawme** [comment]
> My preferred approach is to submit a PR per tool per module e.g. pass cmake tidy over common open PR

OK, I backed out the `clang-format` changes, and the LOC count is down to a more reasonable number. I'll open PRs for the `clang-format` standardization.

File: .github/workflows/tidy-modern.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
+# This is a basic workflow to help you get started with Actions
```
Comment (kunaltyagi) [inline .github/workflows/tidy-modern.yml:1]: @shrijitsingh99 Apologies for pinging you again. Thanks in advance

File: CMakeLists.txt [hunk 4]
```diff
@@ -8,6 +8,7 @@ if(POLICY CMP0074)
 endif()
 
 # Set target C++ standard and required compiler features
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
```
Comment (SergioRAgostinho) [inline CMakeLists.txt:11]: Can we move this to a line 31/32, after the project is set. Also add a comment indicating what forced us to turn it on.

File: common/include/pcl/common/fft/kiss_fft.h [hunk 5]
```diff
@@ -1,9 +1,9 @@
 #pragma once
 
-#include <stdlib.h>
-#include <stdio.h>
-#include <math.h>
-#include <string.h>
+#include <stdlib.h> // NOLINT
+#include <stdio.h>  // NOLINT
+#include <math.h>   // NOLINT
+#include <string.h> // NOLINT
```
Comment (SergioRAgostinho) [inline common/include/pcl/common/fft/kiss_fft.h:6]: kiss_fft is C header. It's probably best to apply no lint to places which include kiss_fft.h

File: common/include/pcl/impl/cloud_iterator.hpp [hunk 6]
```diff
@@ -197,8 +196,7 @@ namespace pcl
       }
 
       ~DefaultConstIterator ()
-      {
-      }
+      override = default;
```
Comment (SergioRAgostinho) [inline common/include/pcl/impl/cloud_iterator.hpp:199]: I would rather delete it entirely. Does clang-tidy provide options for that?

File: common/include/pcl/impl/cloud_iterator.hpp [hunk 7]
```diff
@@ -272,7 +270,7 @@ namespace pcl
       {
       }
 
-      ~ConstIteratorIdx () {}
+      ~ConstIteratorIdx () override = default;
```
Comment (SergioRAgostinho) [inline common/include/pcl/impl/cloud_iterator.hpp:273]: same here.

File: common/include/pcl/point_representation.h [hunk 8]
```diff
@@ -196,7 +196,7 @@ namespace pcl
         trivial_ = true;
       }
 
-      ~DefaultPointRepresentation () {}
+      ~DefaultPointRepresentation () override = default;
```
Comment (SergioRAgostinho) [inline common/include/pcl/point_representation.h:199]: here

File: common/include/pcl/common/fft/kiss_fft.h (hunk 5 above)
Comment (gnawme) [reply common/include/pcl/common/fft/kiss_fft.h:6]: Unless `// NOLINT` is on these lines, `clang-tidy` tries to change these to C++ headers (`<cstdlib>`, etc)

File: common/include/pcl/impl/cloud_iterator.hpp (hunk 6 above)
Comment (gnawme) [reply common/include/pcl/impl/cloud_iterator.hpp:199]: Not explicitly, I'll change it to `= delete`

Comment (SergioRAgostinho) [reply common/include/pcl/impl/cloud_iterator.hpp:199]: I meant deleting the lines of code which explicitly define the destructor.

Comment (gnawme) [reply common/include/pcl/impl/cloud_iterator.hpp:199]: I'm not sure what you're asking, but the diff shows that the `{}` was deleted and replaced by `= default`

File: common/include/pcl/impl/cloud_iterator.hpp [hunk 9]
```diff
@@ -56,9 +56,7 @@ namespace pcl
       {
       }
 
-      ~DefaultIterator ()
-      {
-      }
+      ~DefaultIterator () = default;
```
Comment (SergioRAgostinho) [inline common/include/pcl/impl/cloud_iterator.hpp:59]:
```suggestion
```

File: common/include/pcl/impl/cloud_iterator.hpp [hunk 10]
```diff
@@ -130,7 +128,7 @@ namespace pcl
       {
       }
 
-      virtual ~IteratorIdx () {}
+      virtual ~IteratorIdx () = default;
```
Comment (SergioRAgostinho) [inline common/include/pcl/impl/cloud_iterator.hpp:131]:
```suggestion
```

File: common/include/pcl/impl/cloud_iterator.hpp [hunk 11]
```diff
@@ -196,9 +194,7 @@ namespace pcl
       {
       }
 
-      ~DefaultConstIterator ()
-      {
-      }
+      ~DefaultConstIterator () = delete;
```
Comment (SergioRAgostinho) [inline common/include/pcl/impl/cloud_iterator.hpp:197]:
```suggestion
```

File: .github/workflows/tidy-modern.yml (hunk 3 above)
Comment (shrijitsingh99) [reply .github/workflows/tidy-modern.yml]: No issues.
The current commit has an invalid syntax. 
The previous version has the correct syntax but had some errors.

Something on the lines of this should work (I haven't tested this out).

```yaml
name: tidy-modern

on:
  push:
  pull_request:

jobs:
  tidy:
    runs-on: ubuntu-latest
    
    steps:
    - uses: actions/checkout@v2

    - name: Run clang-tidy
      run: |
          ./.github/workflows/clang-tidy-modern.sh
          git diff > tidy.patch
          cat tidy.patch
          [ ! -s formatting.patch ]
 
```

File: common/include/pcl/impl/cloud_iterator.hpp (hunk 9 above)
Comment (gnawme) [reply common/include/pcl/impl/cloud_iterator.hpp:59]: OK, gotcha.

File: .github/workflows/tidy-modern.yml (hunk 3 above)
Comment (gnawme) [reply .github/workflows/tidy-modern.yml]: Looks clean and simple, I'll revise and give it a try.

Comment (gnawme) [reply .github/workflows/tidy-modern.yml]: As I thought might happen:
```
./.github/workflows/clang-tidy-modern.sh: line 2: run-clang-tidy: command not found

```
@shrijitsingh99 I guess this will require setting up a containerized environment that install `clang-tidy`?

Comment (shrijitsingh99) [reply .github/workflows/tidy-modern.yml]: No need for docker currently.
This should work:

```yaml

name​: ​tidy-modern​

​on​:
  ​push​:
  ​pull_request​:

​jobs​:
  ​tidy​:
    ​runs-on​: ​ubuntu-latest​
    
    ​steps​:
    - ​uses​: ​actions/checkout@v2​

    - ​name​: ​Run clang-tidy​
      ​run​: ​|​
          apt update
          apt install clang-tidy
​          ./.github/workflows/clang-tidy-modern.sh​
​          git diff > tidy.patch​
​          cat tidy.patch​
​          [ ! -s formatting.patch ]​
​
```

Comment (gnawme) [reply .github/workflows/tidy-modern.yml]: @shrijitsingh99 @kunaltyagi Thanks, it finally worked. Here's the [log](https://github.com/gnawme/pcl/actions/runs/170646180).

File: .github/workflows/tidy-modern.yml [hunk 12]
```diff
@@ -0,0 +1,26 @@
+name: tidy-modern
+
+on:
+  push:
+  pull_request:
+
+jobs:
+  tidy:
+    runs-on: ubuntu-latest
```
Comment (kunaltyagi) [inline .github/workflows/tidy-modern.yml:9]: If you choose a docker image of PCL, you don't need to install dependencies (I think you'd need to add clang-tools in the docker image)

File: features/include/pcl/features/3dsc.h [hunk 13]
```diff
@@ -119,7 +119,7 @@ namespace pcl
           rng_.seed (12345u);
       }
 
-      ~ShapeContext3DEstimation() {}
+      ~ShapeContext3DEstimation() override = default;
```
Comment (SergioRAgostinho) [inline features/include/pcl/features/3dsc.h:122]: and this one

File: features/include/pcl/features/board.h [hunk 14]
```diff
@@ -75,7 +75,7 @@ namespace pcl
       }
       
       /** \brief Empty destructor */
-      ~BOARDLocalReferenceFrameEstimation () {}
+      ~BOARDLocalReferenceFrameEstimation () override = default;
```
Comment (SergioRAgostinho) [inline features/include/pcl/features/board.h:78]: and this one

File: features/include/pcl/features/don.h [hunk 15]
```diff
@@ -86,7 +86,7 @@ namespace pcl
         feature_name_ = "DifferenceOfNormalsEstimation";
       }
 
-      ~DifferenceOfNormalsEstimation ()
+      ~DifferenceOfNormalsEstimation () override
       {
         //
       }
```
Comment (SergioRAgostinho) [inline features/include/pcl/features/don.h:92]: and this one

File: features/include/pcl/features/impl/linear_least_squares_normal.hpp [hunk 16]
```diff
@@ -45,8 +45,7 @@
 //////////////////////////////////////////////////////////////////////////////////////////////
 template <typename PointInT, typename PointOutT>
 pcl::LinearLeastSquaresNormalEstimation<PointInT, PointOutT>::~LinearLeastSquaresNormalEstimation ()
-{
-}
+= default;
```
Comment (SergioRAgostinho) [inline features/include/pcl/features/impl/linear_least_squares_normal.hpp:48]: You might want to review your own PR and remomve all these default destructors. I

File: features/include/pcl/features/integral_image2D.h [hunk 17]
```diff
@@ -129,7 +129,7 @@ namespace pcl
 
       /** \brief Destructor */
       virtual
-      ~IntegralImage2D () { }
+      ~IntegralImage2D () = default;
```
Comment (SergioRAgostinho) [inline features/include/pcl/features/integral_image2D.h:132]: This class has no virtual functions so it should no have a virtual destructor. Remove it. 

This is an ABI break.

File: features/include/pcl/features/integral_image2D.h [hunk 18]
```diff
@@ -254,7 +254,7 @@ namespace pcl
...
```
Comment (SergioRAgostinho) [inline features/include/pcl/features/integral_image2D.h:257]: same comment here. ABI break.

File: features/include/pcl/features/multiscale_feature_persistence.h [hunk 19]
```diff
@@ -76,7 +76,7 @@ namespace pcl
       MultiscaleFeaturePersistence ();
       
       /** \brief Empty destructor */
-      ~MultiscaleFeaturePersistence () {}
+      ~MultiscaleFeaturePersistence () override = default;
```
Comment (SergioRAgostinho) [inline features/include/pcl/features/multiscale_feature_persistence.h:79]: remove

File: .github/workflows/tidy-modern.yml (hunk 12 above)
Comment (gnawme) [reply .github/workflows/tidy-modern.yml:9]: OK. @shrijitsingh99, is this as simple as adding a `run: docker run ./.dev/docker/env/Dockerfile` to the action before the other commands?

File: common/src/feature_histogram.cpp [hunk 20]
```diff
@@ -67,9 +67,7 @@ pcl::FeatureHistogram::FeatureHistogram (std::size_t const number_of_bins,
 }
 
 pcl::FeatureHistogram::~FeatureHistogram ()
-{
-  
-}
+= default;
```
[the remark this answers is not in the document]
Comment (gnawme) [reply common/src/feature_histogram.cpp:70]: Causes `undefined reference to `vtable for pcl::FeatureHistogram' if I do

File: common/src/range_image_planar.cpp [hunk 21]
```diff
@@ -51,8 +51,7 @@ namespace pcl
 
   /////////////////////////////////////////////////////////////////////////
   RangeImagePlanar::~RangeImagePlanar ()
-  {
-  }
+  = default;
```
[the remark this answers is not in the document]
Comment (gnawme) [reply common/src/range_image_planar.cpp:54]: Causes `undefined reference to `vtable for pcl::RangeImagePlanar'` if I do

File: common/src/feature_histogram.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (SergioRAgostinho) [reply common/src/feature_histogram.cpp:70]: Did you remove the declaration in the .h file?

File: common/src/range_image_planar.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (SergioRAgostinho) [reply common/src/range_image_planar.cpp:54]: Did you remove the declaration in the .h file?

File: common/src/feature_histogram.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (gnawme) [reply common/src/feature_histogram.cpp:70]: Sorry, didn't save .h before compiling.

File: .github/workflows/clang-tidy-modern.sh [hunk 22]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+run-clang-tidy -header-filter='.*' -checks='-*,modernize-deprecated-headers,modernize-redundant-void-arg,modernize-replace-random-shuffle,modernize-use-equals-default,modernize-use-equals-delete,modernize-use-nullptr,modernize-use-override,modernize-use-using' -fix
```
[the remark this answers is not in the document]
Comment (gnawme) [reply .github/workflows/clang-tidy-modern.sh:2]: Migrated the checks to `.clang-tidy`, will test on next push.

File: features/include/pcl/features/impl/linear_least_squares_normal.hpp (hunk 16 above)
Comment (gnawme) [reply features/include/pcl/features/impl/linear_least_squares_normal.hpp:48]: There were a bunch in the code that were made visible by being changed to `= default`.

What criteria are you using to decide when you want the compiler to create one implicitly?

Comment (gnawme) [reply features/include/pcl/features/impl/linear_least_squares_normal.hpp]: I suppose all of the ones tagged `Empty destructor` would fit...

Comment (SergioRAgostinho) [reply features/include/pcl/features/impl/linear_least_squares_normal.hpp]: With destructors I use the following policy:
- if your class defines virtual methods which it defines (literally defining, not overriding), I explicitly default it. The compiler throws warnings telling if of your class has virtual methods but the destructor is not virtual.
- If a class has empty / default destructors I just remove those lines of code. If the destructor is not implicitly generated you'll get a compiler error, so there's no chance you'll mess up.


For constructors:
- I remove all default constructors unless there's a user specified (non-default) constructor.

File: .github/workflows/tidy-modern.yml (hunk 12 above)
Comment (shrijitsingh99) [reply .github/workflows/tidy-modern.yml:9]: No need for that. Actions already provides  a nice interface to pull and run docker.
This is all you need to do.
```suggestion
    runs-on: ubuntu-latest
    container:
      image: pointcloudlibrary/env:18.04
```

Comment (gnawme) [reply .github/workflows/tidy-modern.yml:9]: Nice, I'll give it a try, thank you.

File: features/include/pcl/features/impl/linear_least_squares_normal.hpp (hunk 16 above)
Comment (gnawme) [reply features/include/pcl/features/impl/linear_least_squares_normal.hpp]: Thanks, sounds like a good focus for another, separate, PR.

File: .github/workflows/tidy-modern.yml (hunk 12 above)
Comment (gnawme) [reply .github/workflows/tidy-modern.yml:9]: @shrijitsingh99 Well, now I get this message:

```
To create a local Git repository instead, add Git 2.18 or higher to the PATH
```
I thought that the checkout action checked out the branch that triggered the action but I guess not.

Is there an action to do that?

Comment (shrijitsingh99) [reply .github/workflows/tidy-modern.yml:9]: It does check it out for you, but AFAIK git will not be available to you inside the docker container, you will have to install it manually.

Comment (gnawme) [reply .github/workflows/tidy-modern.yml:9]: By the time I install `git`, Docker has downloaded the repo via REST. Here's what I need to do:
```
The repository will be downloaded using the GitHub REST API
To create a local Git repository instead, add Git 2.18 or higher to the PATH
```
Where do I modify the Docker `PATH`?

File: gpu/people/src/cuda/nvidia/NCV.hpp [hunk 23]
```diff
@@ -430,7 +430,7 @@ class NCV_EXPORTS NCVMemStackAllocator : public INCVMemAllocator
 
     explicit NCVMemStackAllocator(Ncv32u alignment);
     NCVMemStackAllocator(NCVMemoryType memT, std::size_t capacity, Ncv32u alignment, void *reusePtr=nullptr);
-    ~NCVMemStackAllocator() override;
+    virtual ~NCVMemStackAllocator();
```
Comment (kunaltyagi) [inline gpu/people/src/cuda/nvidia/NCV.hpp]: From the [official docs](https://en.cppreference.com/w/cpp/language/override), `override` is a much better choice than `virtual`
> `override` ensures that the function is virtual **and** is overriding a virtual function from a base class
(emphasis mine)

Comment (gnawme) [reply gpu/people/src/cuda/nvidia/NCV.hpp]: I agree, but these were throwing errors saying something like "marked `override` but does not override".

I'll double-check.

Comment (gnawme) [reply gpu/people/src/cuda/nvidia/NCV.hpp]: OK, the `INCVMemAllocator` abstract base class has a pure virtual destructor, so there is no implementation to override, hence the compilation error when I declare the derived class destructors `override`.

Normally, you only make the destructor pure virtual if the ABC has no other pure virtual methods (which seems like a weird use case).

File: .github/workflows/clang-tidy-modern.sh (hunk 22 above)
[the remark this answers is not in the document]
Comment (gnawme) [reply .github/workflows/clang-tidy-modern.sh:2]: Nope, `run-clang-tidy` doesn't seem to use/find `.clang-tidy`

File: gpu/people/src/cuda/nvidia/NCV.hpp (hunk 23 above)
Comment (gnawme) [reply gpu/people/src/cuda/nvidia/NCV.hpp]: Everything else in `INCVMemAllocator` is pure virtual, I can declare its destructor `virtual ~INCVMemAllocator() = default` and declare the derived class destructors `override`

File: .github/workflows/tidy-modern.yml (hunk 12 above)
Comment (gnawme) [reply .github/workflows/tidy-modern.yml:9]: NVM, I got it sorted, but the 18.04 build is building a different set of files than my 20.04 environment. @kunaltyagi why should that be?

File: gpu/people/src/cuda/nvidia/NCV.hpp (hunk 23 above)
Comment (kunaltyagi) [reply gpu/people/src/cuda/nvidia/NCV.hpp]: I didn't take a look at the code, but I have a handy godbolt: https://godbolt.org/z/8f99dG

**Comment by SergioRAgostinho** [comment]
Split the PR.

**Comment by gnawme** [comment]
> Split the PR.

OK, opened #4283 to deal with empty destructors, this PR reverted to focus on the `clang-tidy` changes.

File: .github/workflows/clang-tidy-modern.sh (hunk 22 above)
[the remark this answers is not in the document]
Comment (kunaltyagi) [reply .github/workflows/clang-tidy-modern.sh:2]: See L199 of https://clang.llvm.org/extra/doxygen/run-clang-tidy_8py_source.html

[the remark this answers is not in the document]
Comment (gnawme) [reply .github/workflows/clang-tidy-modern.sh:2]: Yes, I tried that. It throws an error complaining about an unknown option `-config`. Trying things out locally...

[the remark this answers is not in the document]
Comment (gnawme) [reply .github/workflows/clang-tidy-modern.sh:2]: Weird, `run-clang-tidy -config=''` works on the command line...

[the remark this answers is not in the document]
Comment (gnawme) [reply .github/workflows/clang-tidy-modern.sh]: OK, after jumping through a few hoops, I got `run-clang-tidy` to honor the `.clang-tidy` config.

File: features/include/pcl/features/impl/linear_least_squares_normal.hpp (hunk 16 above)
Comment (gnawme) [reply features/include/pcl/features/impl/linear_least_squares_normal.hpp]: Split off as a separate task for #4283

File: common/include/pcl/common/fft/kiss_fft.h (hunk 5 above)
Comment (kunaltyagi) [reply common/include/pcl/common/fft/kiss_fft.h:6]: I think there's a header-filter regex

File: .github/workflows/clang-tidy-modern.sh [hunk 24]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+run-clang-tidy-10.py -config='' -header-filter='.*' -fix
```
[the remark this answers is not in the document]
Comment (gnawme) [reply .github/workflows/clang-tidy-modern.sh]: Added a terminating newline, it's not showing up in this diff view for some reason

**Comment by gnawme** [comment]
Replaced by #4560, which is taking a phased approach to applying `clang-tidy`
